modules/openthread/platform/diag.c

```diff
@@ -68,3 +71,157 @@ void otPlatDiagAlarmCallback(otInstance *aInstance)
 {
 	ARG_UNUSED(aInstance);
 }
+
+/*
+ * To enable gpio diag commands, in Devicetree create `openthread` node in `/options/` path
+ * with `compatible = "openthread,config"` property and `diag-gpios` property,
+ * which should contain array of GPIO pin's configuration properties containing controller phandles,
+ * pin numbers and pin flags. e.g:
+ *
+ * options {
+ *		openthread {
+ *			compatible = "openthread,config";
+ *			diag-gpios = <&gpio0 0 GPIO_ACTIVE_HIGH>,
+ *				     <&gpio1 0 GPIO_ACTIVE_LOW>;
+ *		};
+ *	};
+ *
+ * To enable reading current gpio pin mode, define
+ * `CONFIG_GPIO_GET_DIRECTION` in prj.conf.
+ *
+ * Note: `<gpio>` in `diag gpio` commands is an index of diag-gpios array. For example shown above,
+ * `ot diag gpio mode 0` will return current mode of pin nmb 0 controlled by `gpio0` controller.
+ */
+#if DT_HAS_COMPAT_STATUS_OKAY(openthread_config) && \
+	DT_NODE_HAS_PROP(DT_COMPAT_GET_ANY_STATUS_OKAY(openthread_config), diag_gpios)
+
+static const struct gpio_dt_spec gpio_spec[] = {
+	DT_FOREACH_PROP_ELEM_SEP(DT_COMPAT_GET_ANY_STATUS_OKAY(openthread_config),
+				 diag_gpios, GPIO_DT_SPEC_GET_BY_IDX, (,))};
+
+static otError gpio_get_spec(uint32_t gpio_idx, const struct gpio_dt_spec **spec)
+{
+	otError error = OT_ERROR_NONE;
+
+	otEXPECT_ACTION(gpio_idx < otARRAY_LENGTH(gpio_spec), error = OT_ERROR_INVALID_ARGS);
+
+	*spec = &gpio_spec[gpio_idx];
+
+	otEXPECT_ACTION(otPlatDiagModeGet(), error = OT_ERROR_INVALID_STATE);
+	otEXPECT_ACTION(device_is_ready((*spec)->port), error = OT_ERROR_INVALID_ARGS);
+
+	const struct gpio_driver_config *const cfg =
+		(const struct gpio_driver_config *)((*spec)->port->config);
+
+	otEXPECT_ACTION((cfg->port_pin_mask & (gpio_port_pins_t)BIT((*spec)->pin)) != 0U,
+			error = OT_ERROR_INVALID_ARGS);
+
+exit:
+	return error;
+}
+
+otError otPlatDiagGpioSet(uint32_t aGpio, bool aValue)
+{
+	const struct gpio_dt_spec *spec;
+	otError error;
+	int rv;
+
+	error = gpio_get_spec(aGpio, &spec);
+
+	otEXPECT(error == OT_ERROR_NONE);
+
+#if defined(CONFIG_GPIO_GET_DIRECTION)
+	rv = gpio_pin_is_output_dt(spec);
+	otEXPECT_ACTION(rv == 1, error = OT_ERROR_FAILED);
```

Comment:
```suggestion
	otEXPECT_ACTION(rv == 1, error = OT_ERROR_INVALID_STATE);
```
to comply with https://github.com/openthread/openthread/pull/8538